enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md

```diff
@@ -0,0 +1,378 @@
+---
+title: share-secrets-and-configmaps-across-namespaces-via-a-csi-driver
+
+authors:
+  - "@gabemontero"
+
+reviewers:
+  - "@bparees"
+  - "@adambkaplan"
+  - "@derekwaynecarr"
+  - "@gnufied"
+  
+approvers:
+  - "@bparees"
+  - "@adambkaplan"
+  - "@derekwaynecarr"
+  - "@gnufied"
+  
+creation-date: 2020-03-17
+
+last-updated: 2020-03-30
+
+<!-- status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced -->
+status: provisional
+
+
+  
+---
+
+# Share Secrets And ConfigMaps Across Namespaces via a CSI Driver
+
+
+## Release Signoff Checklist
+
+- [/] Enhancement is `implementable`
+- [/] Design details are appropriately documented from clear requirements
+- [/] Test plan is defined
+- [/] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+> 1. Is a CSI `SecretProviderClass` based implementation recommended?  Received some varying guidance.  Also, per PR
+>review from @jsafrane, of which I can only concur based on my examination, the example at https://github.com/kubernetes-sigs/secrets-store-csi-driver
+>does not appear to handle refreshing of the data after initial provisioning.  So it does not appear to be something we
+>can use as is.  But do we want to take on adding such a feature upstream to that repo? **This question most
+>likely can be deferred until implementation.**
+> 2. What will be the precise means of interacting with the host filesystem.  The hostpath plugin actually uses the 
+>simple golang `os.MkdirAll` for example when creating mount access based volumes.  Can we do that in all our supported
+>scenarios?  We've also received suggestions around tmpfs or read only bind-mounts of a directory from the CSI driver.
+>**This question most likely can be deferred until implementation.**
+> 3. SELinux implications will need to be properly understood.  Minimally, a consuming Pod needs the correct SELinux 
+>context and fs permissions with bind-mounts for example.
+>**This question most likely can be deferred until implementation.**
+> 4. The EP document in https://github.com/openshift/enhancements/pull/169 needs to be reconciled with this document in
+>some form or fashion.
+
+## Summary
+
+This proposal will describe a [CSI driver](https://github.com/container-storage-interface/spec/blob/master/spec.md) based
+means for OpenShift cluster administrators to share `Secrets` / `ConfigMaps` defined in one namespace with
+other namesapces. 
+
+Any owners/authors/users of Pods in those namespaces can opt into the consumption of the content of those shared `Secrets` / 
+ `ConfigMaps` during their Pod's execution.
+ 
+The precise means of opting into consuming the shared content are outside the scope of this proposal.
```

Comment:
this concerns me.  Part of the problem is distributing the set of secret and configmap names that can be consumed.

Reply:
That is what https://github.com/openshift/enhancements/pull/214 is for (that EP needs a retitling, i think.  That EP is intended to cover how we control which pods can consume which secrets, and auto-configuring pods(adding the volume) to do so in certain cases.

This EP is really just to implement the dependent bit of work of creating a CSI driver that allows any pod to mount any secret/configmap.  It is not about which pods are allowed to do it, or automation around configuring pods w/ certain secrets/configmaps.

Reply:
yeah there were pros and cons with splitting out the CSI driver bits from #214 as we decided to do

and a con is for folks coming in mid stream like @deads2k is here don't have the history

don't know if there are any guidelines upstream wrt how separate EPs are allowed to coordinate

it is a slippery slope, but based on @deads2k `ProjectedResource` suggestion below, I'm thinking at least part of the opt in piece neeeds to be here in this EP

I'll take a stab at it, change this line to reflect the new dividing line between this EP and #214, 
and we can then see what form of consensus we have